Properly parse and handle DBS client errors

vkuznet opened the pull request:
Fixes #12229 

#### Status
ready

#### Description
Properly collect all DBS error codes and provide new method to represent them to the client. The client code can request for DBS error codes and interpret them accordingly. Adjust DBSUploadPoller to use DBS error codes and perform necessary logic to use DBS error message:
- added check for block existence before inserting it to DBS
- this eliminates hard-coded codes in DBSUploadPoller
- and generalize usage of DBSError message for reporting
- the DBSError now returns code of logic operation which we look-up from DBS errors dict, obtained from`/errors` end-point, and rely on DBS message to show DBS error flow.

#### Is it backward compatible (if not, which system it affects?)
NO, the DBS error codes were re-grouped on DBS server side to provide the following ranges:
- 1xx codes are allocated for generic errors, e.g. `DatabaseErrorCode` or `ReaderErrorCode`
- 2xx codes are allocated for logic errors, e.g. `BlockAlreadyExist` or `PrimaryDatasetDoesNotExist`
- 3xx codes are allocated for database insert operations, e.g. `InsertBlockErrorCode` or `InsertDatasetAccessTypeErrorCode`
- 4xx codes are allocated for transient errors, like missing data in DBS, e.g. `GetDataTierIDErrorCode` when we look-up foreign key for provided data tier, and if data tier is not present in DBS this error will be raised. Please note that DBS message should provide further details about which data-tier value was not present the codes themselves are not aware of specific values stored in DBS.
- 5xx codes are allocated for update, remove, clean-up operations like `UpdateBlockErrorCode` or `CancelMigrationErrorCode`

Therefore, in order to apply these changes we first need to upgrade dbs server and then can use this changes in a client.

#### Further improvements
It is possible to improve further how WM client will treat DBS error codes. For instance we can use provided ranges to improve and generalize return message. For instance, once we got DBS error code we canexplicitly say if it is insert or transient error. But new DBS error message will provide these details too.

The way how DBS errors are displayed in WM client code is shown in this https://github.com/dmwm/WMCore/pull/12312#issuecomment-2840313254

#### Related PRs
Here is complete list of DBS server related PRs to improve DBS error handling in client:
- [dmwm/dbs2go#121](https://github.com/dmwm/dbs2go/pull/121) fix parent file error code when parent of lfn is not found
- [dmwm/dbs2go#122](https://github.com/dmwm/dbs2go/pull/122) improve DBS error output (make it human readable) for nested errors
- [dmwm/dbs2go#124](https://github.com/dmwm/dbs2go/pull/124) CChange generic error codes to expilcit ones by addressing commits, looks-up and insert of different parts of bulkblock API
- [dmwm/dbs2go#129](https://github.com/dmwm/dbs2go/pull/129) New set of DBS Error codes with explicit content about all failures/errors
- [dmwm/dbs2go#130](https://github.com/dmwm/dbs2go/pull/130) add explanation of new DBS error codes
- [dmwm/dbs2go#131](https://github.com/dmwm/dbs2go/pull/131) to provide `/errors` end point and allow discovery of DBS error codes
- [dmwm/dbs2go#137](https://github.com/dmwm/dbs2go/pull/137) use explicit codes

**Examples:**
Below we provide examples of concrete DBSError and output of `/errors` end-point used to provide meaning of all DBS error codes.

Here is concrete example of failed DBS server API:
```
scurl -s "https://xxxx.cern.ch:port/dbs/int/global/DBSReader/datatiers?dataset=bla" | jq
[
  {
    "error": {
      "reason": "\nDBSError\n   Code: 118\n   Description: DBS invalid parameter for the DBS API or validation error\n   Function: dbs.validator.Check\n   Message: unable to match 'dataset' value 'bla'\n   Reason: invalid parameter(s)\n",
      "message": "not str type",
      "function": "dbs.Validate",
      "code": 113,
      "stacktrace": "\ngoroutine 91170 [running]:\ngithub.com/dmwm/dbs2go/dbs.Error({0xbd67a0?, 0xc0006a7200?}, 0x71, {0xae5205, 0xc}, {0xae5211, 0xc})\n\t/go/src/github.com/vkuznet/dbs2go/dbs/errors.go:383 +0x99\ngithub.com/dmwm/dbs2go/dbs.Validate(0xc00006e5e8?)\n\t/go/src/github.com/vkuznet/dbs2go/dbs/validator.go:337 +0x394\ngithub.com/dmwm/dbs2go/web.validateMiddleware.func1({0xbda670, 0xc000646168}, 0xc000310640)\n\t/go/src/github.com/vkuznet/dbs2go/web/middlewares.go:91 +0x7b\nnet/http.HandlerFunc.ServeHTTP(0x76925b?, {0xbda670?, 0xc000646168?}, 0x0?)\n\t/usr/local/go/src/net/http/server.go:2294 +0x29\ngithub.com/dmwm/dbs2go/web.authMiddleware.func1({0xbda670, 0xc000646168}, 0xc000310640)\n\t/go/src/github.com/vkuznet/dbs2go/web/middlewares.go:79 +0x343\nnet/http.HandlerFunc.ServeHTTP(0x10?, {0xbda670?, 0xc000646168?}, 0xc00006e848?)\n\t/usr/local/go/src/net/http/server.go:2294 +0x29\ngithub.com/vkuznet/auth-proxy-server/logging.LoggingMiddleware.func1({0xbda4d8, 0xc00055c0e0}, 0xc000310640)\n\t/go/pkg/mod/github.com/vkuznet/auth-proxy-server/lo"
    },
    "http": {
      "method": "GET",
      "code": 400,
      "timestamp": "2025-05-06 16:42:17.020682622 +0000 UTC m=+61395.357610561",
      "path": "/dbs/int/global/DBSReader/datatiers?dataset=bla",
      "user_agent": "curl/8.13.0",
      "x_forwarded_host": "xxxx.cern.ch",
      "x_forwarded_for": "xxx.xxx.xxx.xxx",
      "remote_addr": "xxx.xxx.xxx.xxx:50990"
    },
    "exception": 400,
    "type": "HTTPError",
    "message": "\nDBSError\n   Code: 113\n   Description: DBS validation error, e.g. input parameter does not match lexicon rules\n   Function: dbs.Validate\n   Message: not str type\n   Reason: \nDBSError\n   Code: 118\n   Description: DBS invalid parameter for the DBS API or validation error\n   Function: dbs.validator.Check\n   Message: unable to match 'dataset' value 'bla'\n   Reason: invalid parameter(s)\n\n"
  }
]
```
And, here is (partial) output of new `/errors` DBS end-point:
```
scurl -s "https://xxxx.cern.ch:port/dbs/int/global/DBSReader/errors" | jq |
 head
[
  {
    "code": 100,
    "meaning": "Generic DBS error"
  },
  {
    "code": 101,
    "meaning": "DBS DB error"
  },
  {
...
]
```
- `error.code` defines high level DBS error code produced by DBS server upon the error
- `error.message` defines DBS message when this `code` was happen
- `error.function` defines DBS function where this happen
- `error.reason` defines reason why this error happen (it can be recursive calls through DBS codebase up to a database but not necessary)

Also, the DBS error provides `message` at high level JSON output which includes `error.code`, followed by `error.message` and finally followed by `error.reason` as newline formatted string.

The DBS `/errors` end-point provides description (meaning) of all existing DBS error codes. Therefore, usage of `error.code` and `meaning` is sufficient to compose friendly report in a client. More verbose output can be provided via utilization all of the above information or just using `message` from DBS error JSON response.

#### External dependencies / deployment changes

vkuznet committed:
Add getCodes method to fetch all dbs2go codes from its error JSON response

vkuznet committed:
Add unit tests

mapellidario reviewed (approved):
If I understand correctly, these changes make sure that when we fail to insert a block in dbs, we keep track in the agent of all the errors that have been encountered inside dbs, to make sure that we always have the root cause for the error on our side.

If this is the case, then i approve, looks good to me

amaltaro reviewed (changes requested):
Before going with a full review, I'd rather have those commits properly squashed first - as every now and then I observe changes getting lost during the squashing process.

In addition, I think it is important to first move with the server changes (dbs2go). Once testbed is running the new dbs2go version, please test these changes in an agent with a **representative** use scenario.

Lastly, I think our agreement was that either `error.reason` or `message` would have a clear and friendly error message (with the actual root error). Did I misunderstand our discussion? Why do we have to have string parsing of the error served by the server?

vkuznet commented:
@amaltaro please note that reported unit test failures are not related to this PR changes.

anpicci commented:
test this please

amaltaro commented on src/python/WMCore/Services/DBS/DBSErrors.py:
```diff
@@ -96,6 +96,17 @@ def getServerCode(self):
         return 0
 
+    def getCodes(self):
+        """
```
https://github.com/dmwm/WMCore/blob/master/CONTRIBUTING.rst#project-docstrings-best-practices

todor-ivanov replied:
I do not get the comment here.... @amaltaro, what is not followed in this docstring by the book?

anpicci replied:
I would leave the description as is, given that the function is not parsing and then separately putting the results in a list, but it's creating a list making the associations between the error messages and the error codes

amaltaro replied:
I think the implementation of the function/method does not influence the structure of the docstring.
But if people want to update the contribution guidelines with a different recommendation, let us do it. 
I am just trying to seek consistency in the project.

amaltaro reviewed (changes requested):
As mentioned in my previous review, I think we should not be traversing all the errors from the server in our client layer, instead we should use the final and root cause error that the server will be providing. I also left a comment along the code.

Before we converge and merge this PR, I think we need to:
a) upgrade `dbs2go` in testbed
b) update this PR to use the actual precise error (`message` or `reason` field?)
c) apply the candidate PR/patch to an agent
d) test this with a representative scenario.

anpicci commented on src/python/WMComponent/DBS3Buffer/DBSUploadPoller.py:
```diff
@@ -99,20 +99,20 @@ def uploadWorker(workInput, results, dbsUrl, gzipEncoding=False):
-                msg = f"Error trying to process block {name} through DBS. Details: {msg}"
-                logging.error(msg)
-                results.put({'name': name, 'success': "error", 'error': msg})
+            for srvCode in dbsError.getCodes():
```
@amaltaro , I suppose this comment is outdated after the MM discussion, right?

vkuznet commented:
I addressed @amaltaro concern about "friendly" messages via the following PR: https://github.com/dmwm/dbs2go/pull/129 In particular, I modified **520 (!!!) errors** in DBS server code and move away from generic errors to concrete ones which address friendliness of returned DBS error code error and messages. I'll provide additional changes in this PR to address backward compatibility and avoid explicit reliance on DBS error code values. Stay tuned.

vkuznet commented:
@amaltaro , @anpicci , @todor-ivanov , @mapellidario I request new review of this PR which has new logic based on changes provided by DBS server. Please see description for details about DBS changes. And make review keeping in mind these changes which are required for this PR. Due to re-organization of DBS error codes in logical groups, i.e. generic errors, insert errors, transient errors, this PR breaks backward compatibility since it used hard-coded DBS codes. Moreover, I made changes in WM code to be agnostic to DBS code assignment and also provide further improvements suggestions in description of this PR.

amaltaro commented on src/python/WMComponent/DBS3Buffer/DBSUploadPoller.py:
```diff
@@ -90,29 +90,29 @@ def uploadWorker(workInput, results, dbsUrl, gzipEncoding=False):
         # Do stuff with DBS
         try:
+            # check if block exists in DBS
+            results = dbsApi.listBlocks(block_name=name)
```
Given that most of the blocks succeed insertion in the first attempt, I feel like this check is a waste of resources.
I think I would stick with the current implementation, which marks the block is injected if we get a "Block already inserted" error from the server.
Or another option would be to create a list of blocks that we need to verify existence in the server.

vkuznet replied:
Alan, let's not over optimize where it is not necessary. Checking for block API call is nothing to DBS server, and neither take significant time. Here is benchmark
```
time scurl "https://cmsweb-testbed.cern.ch/dbs/int/global/DBSReader/blocks?block_name=/ZMM_13TeV_TuneCP5-pythia8/RunIIAutumn18DR-SNB_102X_upgrade2018_realistic_v17-v2/AODSIM%2340d9b846-ef36-43db-9094-9367421c5f71"
[
{"block_name":"/ZMM_13TeV_TuneCP5-pythia8/RunIIAutumn18DR-SNB_102X_upgrade2018_realistic_v17-v2/AODSIM#40d9b846-ef36-43db-9094-9367421c5f71"}
]
scurl   0.02s user 0.01s system 7% cpu 0.422 total
```
So, it takes 0.02 seconds from US to EU link.

That said, the new DBS server code properly report 'Blck <bla> already exists', see https://github.com/dmwm/dbs2go/blob/master/dbs/bulkblocks2.go#L456

If you still don't want extra HTTP call to check the block I'll remove it, but I want to remove hard-code code check in WM client as it constrain and lead to potential backward incompatibility with DBS server.

amaltaro replied:
Even though this query is fast, I've seen APS hanging requests for minutes (see JobUpdater timeouts).
That to say that, if we are doing things that are not really required, this can always be a choking point in the algorithm.

Anyhow, I think we can proceed with your implementation, as in this way we can stop looking for specific error codes, as you pointed out.

vkuznet replied:
again, I don't mind to remove it, but I think without details of JobUpdater timeouts I can't say much if your statement is accurate or not. I don't really know what JobUpdater timeout is and which queries it did, and even if it query simple block name. But I provided simple curl call and measure its time. I can do more comprehensive benchmark if necessary with `hey` tool. The block look-up query is fast and should be fast and most likely what you concern with JobUpdater may not be related to it.

I do agree that we should avoid unnecessary things and the reason I introduced this check of block existence since current WM code checks for DBS error and use hard-coded one which I wanted to avoid. The new DBS server codes should properly report if block already exists. My point is that I try to make changes in WM code which will be **backward** compatible with either DBS server, therefore checking for block before injection will ensure that WM code will work with both current and new DBS server. And, because each HTTP call is treated individually in FE there is not hard on algorithm because `listBlocks` time will not be accounted as part of `injectionBulkBlock` API but will only contribute to total time of the loop (even though as I measured it should be negligible).

Finally, if JobUpdater timeouts is real issue I rather prefer address it in separate ticket and do not mix with this development.

vkuznet replied:
For the reference, here is benchmark using `hey` tool with dbs blocks queries using 200 request and 50 concurrent clients against cmsweb-testbed
```
/cvmfs/cms.cern.ch/cmsmon/hey -U ~/dbs-blocks.txt
All requests done.

Summary:
  Total:        0.3109 secs
  Slowest:      0.2897 secs
  Fastest:      0.0042 secs
  Average:      0.0471 secs
  Requests/sec: 643.2064

Status code distribution:
  [200] 200 responses

Response time histogram:
  0.004 [1]     |
  0.033 [121]   |∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
  0.061 [26]    |∎∎∎∎∎∎∎∎∎
  0.090 [7]     |∎∎
  0.118 [15]    |∎∎∎∎∎
  0.147 [13]    |∎∎∎∎
  0.176 [8]     |∎∎∎
  0.204 [5]     |∎∎
  0.233 [0]     |
  0.261 [0]     |
  0.290 [4]     |∎

Latency distribution:
  10% in 0.0055 secs
  25% in 0.0071 secs
  50% in 0.0136 secs
  75% in 0.0700 secs
  90% in 0.1350 secs
  95% in 0.1721 secs
  99% in 0.2894 secs
```
I don't see any delay in block queries and on average it takes 0.033 sec.

amaltaro commented on src/python/WMCore/Services/DBS/DBSErrors.py:
```diff
@@ -111,3 +127,20 @@ def getReason(self):
             return self.data['error']['reason']
         return ""
+
+def DBSErrors(dbsUrl):
```
I would suggest not to mix the REST API call with the class dealing with a response object from the DBS server.

Perhaps a better place for it would be either in DBS3Reader:
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/Services/DBS/DBS3Reader.py
or DBSUtils, or DBSConcurrency... things are pretty much scattered out.

vkuznet replied:
fine I can move it there. To me it is more natural to keep errors with definition to error.

vkuznet replied:
ok, I moved code to DBSUtils.

amaltaro commented on src/python/WMComponent/DBS3Buffer/DBSUploadPoller.py:
```diff
@@ -90,29 +90,29 @@ def uploadWorker(workInput, results, dbsUrl, gzipEncoding=False):
-                msg = f"Error trying to process block {name} through DBS. Details: {msg}"
-                logging.error(msg)
-                results.put({'name': name, 'success': "error", 'error': msg})
+            errors = DBSErrors(dbsUrl)
```
Other than the concern expressed above, how much is the output of DBSErrors supposed to change?
I believe it will be pretty much static, and if so, we could use a caching system and avoid much of the I/O.

vkuznet replied:
Once again I against over optimization where it is not obvious. This API call to DBS server do not even touch its database. The HTTP cal which `DBSErrors` do is negligible, while caching will lead to a possibility that if we upgrade dbs server and it changes code (add new code) that we'll have stale cache which someone will need to take into account. I don't see much need to introduce such cache.

amaltaro replied:
Yes, I agree that using cache always increases the complexity of the code. On the other hand, I feel like we need to adopt efficient design and implementation whenever we see fit, especially as the scale of data will only increase in the next years.

I guess a good compromisse would be to fetch these server errors once this method `uploadWorker` starts, above the `while True`. This way we ensure that we have only one call for each cycle of the component.
What do you think?

vkuznet replied:
ok, I agree , it is good suggestion to move it outside of while loop.

amaltaro reviewed (changes requested):
Thank you for following this up, @vkuznet .
In addition to the comments made along the code, I am fairly uncomfortable with the error code ranges. The reason is that they end up with some significant overlap, for instance: logical errors (2xx) can be raised from insert operations (3xx), which could arise from transient errors as well (4xx).

With that said, I think the idea is great, but in practice it will be hard to trust the code ranges. In my opinion, having a static proper exit code and error description is all that the client needs.

Said that, perhaps we should extend the error codes, but not modify those already deployed in production - as it affects at least production and analysis, for data insertion.

vkuznet commented:
@amaltaro , due to implemented DBS server codes grouping, I can change groups such that we may have backward compatible codes, e.g. the code `128` which currently is used by WM code to check that block already exists. But it does not solve backward compatibility since current WM code uses some set of codes, e.g.
```
            if srvCode == 128:
                # block already exist
                logging.warning("Block %s already exists. Marking it as uploaded.", name)
                results.put({'name': name, 'success': "check"})
            elif srvCode in [132, 133, 134, 135, 136, 137, 138, 139, 140]:
                # racing conditions
                logging.warning("Hit a transient data race condition injecting block %s, %s", name, msg)
                results.put({'name': name, 'success': "error", 'error': msg})
```
and I introduced in DBS server more codes that this block checks. My suggestion to remove hard-coded checks for DBS codes and doing so will provide backward compatibility of WM code with current DBS server and new one where I introduced code grouping.

I do see benefits in DBS error code grouping, but as you said it depends on a view what they address. Someone may argue (as you pointed out) that one group comes from another. The groups can be done in different ways, e.g. by block, file, tier, etc. levels too. For instance groups 2xx, 3xx, and 4xx may be combined since they reflects how we inject data but represents different aspects of it.

Bottom line, I am see more danger in having **re-interpretation** of DBS codes in clients like now we have in DBSUploadPoller, i.e. it reads DBS error code, and return its own error. But I agreed and made huge effort to improve DBS error themselves to be more clear for the clients. That said, the new DBS code returns (as it best knowledge) details of errors and more concise error about where it happen in properly formatted way.

anpicci commented:
Hi @vkuznet, @amaltaro, I believe we can start digesting the new DBS error coding system as a client on the WM side, with the commitment from both sides to interact with each other to improve this coding system, in the event the client would benefit from some further refactoring. What do you think?

amaltaro commented:
I didn't see you asking for extra review of this PR, but I can see a few blockers that need to be addressed:
* Jenkins report is your friend, please use it.
* some unit tests have been marked as deleted - but I suspect you did not mean to do that - which likely means you have a broken import
* some pylint score got degraded, hinting to must-fix changes.

vkuznet commented:
test this please

vkuznet commented:
test this please

vkuznet commented:
@amaltaro , I finished my changes with this PR and requesting now the review. That said, there is one failed unit test, see
https://cmssdt.cern.ch/dmwm-jenkins/job/WMCore-PR-Report/606/testReport/junit/WMComponent_t.DBS3Buffer_t.DBSUpload_t/DBSUploadTest/testCloseSettingsPerWorkflow/

I tried to understand its failure by adding some logging info. The failed `DBSUploadTest.testCloseSettingsPerWorkflow` unit test uses hard-coded dbs url:
[`self.dbsUrl = "https://localhost:1443/dbs/dev/global/DBSWriter"`](https://github.com/dmwm/WMCore/blob/master/test/python/WMComponent_t/DBS3Buffer_t/DBSUpload_t.py#L73) which I have no idea how it is setup and used. Since it is localhost I don't know where it is setup, but it should be available in Jenkins. Moreover, it is unclear which JSON content it is loaded to tests. My recent commit to improve these topics does not reveal anything useful and I can easily roll it back. But I would appreciate if you'll explicitly look into the failed unit test to clarify how dbs should be used. It may or may not be related to changes I made in this PR.

Finally, I did installed this PR on my vocms266 agent via Todor's venv setup. And, with these changes I was able to start all components and agent itself is running fine.

vkuznet commented:
@amaltaro , I finally found issue with unit test and now all tests are passed. Feel free to review this PR again.

vkuznet commented:
I was able to test DBS error client's changes with current (production) version `v00.06.44` of DBS server  and newer version `v00.06.46` of DBS server using cmsweb-testbed. Here are how errors are reported:

##### current version `v00.06.44` (which we run on production):
```
Traceback (most recent call last):
  File "/data/WMAgent.venv3/lib64/python3.9/site-packages/dbs/apis/dbsClient.py", line 474, in __callServer
    self.http_response = method_func(self.url, method, params, data, request_headers)
  File "/data/WMAgent.venv3/lib64/python3.9/site-packages/RestClient/RestApi.py", line 42, in post
    return http_request(self._curl)
  File "/data/WMAgent.venv3/lib64/python3.9/site-packages/RestClient/RequestHandling/HTTPRequest.py", line 62, in __call__
    raise HTTPError(effective_url, http_code, http_response.msg, http_response.raw_header, http_response.body)
RestClient.ErrorHandling.RestClientExceptions.HTTPError: HTTP Error 400:

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/data/WMAgent.venv3/DBSUploadPoller.py", line 78, in uploadWorker
    dbsApi.insertBulkBlock(blockDump=block)
  File "/data/WMAgent.venv3/lib64/python3.9/site-packages/dbs/apis/dbsClient.py", line 647, in insertBulkBlock
    result =  self.__callServer("bulkblocks", data=blockDump, callmethod='POST' )
  File "/data/WMAgent.venv3/lib64/python3.9/site-packages/dbs/apis/dbsClient.py", line 477, in __callServer
    self.__parseForException(http_error)
  File "/data/WMAgent.venv3/lib64/python3.9/site-packages/dbs/apis/dbsClient.py", line 524, in __parseForException
    raise exp
  File "/data/WMAgent.venv3/lib64/python3.9/site-packages/dbs/apis/dbsClient.py", line 522, in __parseForException
    raise HTTPError(http_error.url, data['exception'], data['message'], http_error.header, http_error.body, serverCode(data))
RestClient.ErrorHandling.RestClientExceptions.HTTPError: HTTP Error 400: DBSError Code:139 Description:Not defined Function:dbs.bulkblocks.getDatasetAccesssTypeID Message:fac32f6ca085cd6c22dba0e827c4bc7ec1413c7830d7dc4e673e74934557c51d unable to find dataset_access_type_id for data access type='' Error: nested DBSError Code:110 Description:DBS DB insert record error Function:dbs.GetRecID Message: Error: nested DBSError Code:112 Description:DBS DB laster insert record error, e.g. fail to obtain last inserted ID Function:dbs.datasetaccesstypes.Insert Message: Error: nested DBSError Code:112 Description:DBS DB laster insert record error, e.g. fail to obtain last inserted ID Function:dbs.IncrementSequence Message: Error: nested DBSError Code:103 Description:DBS DB query error, e.g. mailformed SQL statement Function:dbs.IncrementSequences Message: Error: ORA-02289: sequence does not exist

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/data/WMAgent.venv3/DBSUploadPoller.py", line 99, in <module>
    uploadWorker(dbsUrl)
  File "/data/WMAgent.venv3/DBSUploadPoller.py", line 87, in uploadWorker
    msg += f'\nDBSError code: {srvCode}, meaning: {dbsErrors[srvCode]}'
KeyError: 139
```
The error code `139` corresponds to `DatasetAccessTypeDoesNotExist` in old production DBS server code.

##### newer version `v00.06.46` (which we plan to deploy to production):
```
Traceback (most recent call last):
  File "/data/WMAgent.venv3/DBSUploadPoller.py", line 69, in uploadWorker
    dbsApi.insertBulkBlock(blockDump=block)
  File "/data/WMAgent.venv3/lib64/python3.9/site-packages/dbs/apis/dbsClient.py", line 647, in insertBulkBlock
    result =  self.__callServer("bulkblocks", data=blockDump, callmethod='POST' )
  File "/data/WMAgent.venv3/lib64/python3.9/site-packages/dbs/apis/dbsClient.py", line 477, in __callServer
    self.__parseForException(http_error)
  File "/data/WMAgent.venv3/lib64/python3.9/site-packages/dbs/apis/dbsClient.py", line 524, in __parseForException
    raise exp
  File "/data/WMAgent.venv3/lib64/python3.9/site-packages/dbs/apis/dbsClient.py", line 522, in __parseForException
    raise HTTPError(http_error.url, data['exception'], data['message'], http_error.header, http_error.body, serverCode(data))
RestClient.ErrorHandling.RestClientExceptions.HTTPError: HTTP Error 400:
DBSError
   Code: 237
   Description: dataset access type does not exist
   Function: dbs.bulkblocks.getDatasetAccesssTypeID
   Message: fac32f6ca085cd6c22dba0e827c4bc7ec1413c7830d7dc4e673e74934557c51d unable to find dataset_access_type_id for data access type=''
   Reason:
DBSError
   Code: 110
   Description: DBS DB insert record error
   Function: dbs.GetRecID
   Message:
   Reason:
DBSError
   Code: 112
   Description: DBS DB laster insert record error, e.g. fail to obtain last inserted ID
   Function: dbs.datasetaccesstypes.Insert
   Message: unable to increment dataset access types sequence number
   Reason:
DBSError
   Code: 112
   Description: DBS DB laster insert record error, e.g. fail to obtain last inserted ID
   Function: dbs.IncrementSequence
   Message:
   Reason:
DBSError
   Code: 103
   Description: DBS DB query error, e.g. malformed SQL statement
   Function: dbs.IncrementSequences
   Message:
   Reason: ORA-02289: sequence does not exist

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/data/WMAgent.venv3/DBSUploadPoller.py", line 86, in <module>
    uploadWorker(dbsUrl)
  File "/data/WMAgent.venv3/DBSUploadPoller.py", line 77, in uploadWorker
    msg += f'\nDBSError code: {srvCode}, meaning: {dbsErrors[srvCode]}'
KeyError: 237
```
the error code `237` corresponds to `DatasetAccessTypeDoesNotExist` in new production code. As you can see new DBS error is much cleaner and shows full chain of errors starting from high-level one and goes down to actual ORACLE database error via intermediate errors during the chain of calls.

Said that, I found mismatch of DBS server error codes between run-time assigned error code and the ones I put in DBS error code comments. For that I provided new PR https://github.com/dmwm/dbs2go/pull/137 on DBS server side and asked @todor-ivanov to review, merge and deploy it to testbed. Once this PR will be merged I'll another round of integration testing with my agent, and report back results here. The change would be proper code value which now is hard-coded within DBS server instead of incremental one. If you are interested in details of this change please see my comment: https://github.com/dmwm/dbs2go/pull/137#issuecomment-2840184931

amaltaro commented on src/python/WMComponent/DBS3Buffer/DBSUploadPoller.py:
```diff
@@ -90,29 +92,29 @@ def uploadWorker(workInput, results, dbsUrl, gzipEncoding=False):
+            if len(records) == 1 and records[0]['block_name'] == name:
+                # found that we have this block, i.e. no need to insert block anymore
+                logging.info("block %s already exist in DBS", name)
+                continue
```
Before continue, don't we have to declare this block as injected, hence:
```
            results.put({'name': name, 'success': "uploaded"})
```
? Such that the agent updates the block status in the local database. Otherwise this block will keep showing up in the cycle, every time.

vkuznet replied:
yes, you are right, I added entry to the queue.

amaltaro commented on src/python/WMCore/Services/DBS/DBSErrors.py:
```diff
@@ -96,6 +96,20 @@ def getServerCode(self):
             return self.data['error']['code']
         return 0
 
+    def getCodes(self):
```
I don't see this method being used anywhere other than in the unit tests. If this is no longer relevant, I would suggest to actually remove it.
If it is used only inside the test, then I think it would be better to move this implementation out of the `src/` directory (likely in the test/ one).

vkuznet replied:
fine, removed

amaltaro commented on src/python/WMCore/Services/DBS/DBSUtils.py:
```diff
@@ -137,3 +139,28 @@ def urlParams(url):
             rdict[key] = vals[0]
     return rdict
+
+def DBSErrors(dbsUrl):
```
I thought I had seen an unit test for this new method (when I suggested to mock it - after running a real unit test and ensuring it works as expected). Did it get lost during the squash process?

vkuznet replied:
there were no test provided initially, I just add new integration unit test for it.. Regarding mocking itself, I'm not sure where and how to do it. If it is required please elaborate on this request more. For instance, I can write new function in src/python/WMQuality/Emulators/DBSUtils/DBSUtils.py to simulate output of DBS `/errors` end-point (it is a list of dicts) but then I don't know where to use it.

amaltaro reviewed (changes requested):
@vkuznet I left a few comments for your consideration along the code

vkuznet commented:
@amaltaro , I resolved issue after your last review, e.g. removed `getCodes`. The code now only uses high level DBS error as you originally requested. Since I proposed to use all intermediate errors (and for that `getCodes` method was developed) I would like to double check with you that you're ok to only rely on high level DBS error and discard all intermediate ones. I still consider though that logging all DBS errors is beneficial to understand and debug issue with DBS server/request since the amount of DBS errors is very high and I don't know if we cover all possible combinations/conditions when error occur within WM code. But I put tremendous effort into DBS server to ensure that high level error is clear. Please review code and again and make final judgement call about DBS errors (the commits are not squashed and we can put back `getCodes` and its unit tests and use it if necessary).

amaltaro commented:
@vkuznet about printing the whole error stack, how about we only do that when the component log level is set to DEBUG? Hence, we could have it in the component under a `logging.debug()` statement.

For the DBS mocking, AFAIR you have to:
1) patch the appropriate module
2) create the API call signature
3) run a script that will really call DBS Server and it will create a huge JSON file with the content (+ the new content out of DBSErrros API)


But you can find further information under https://cms-wmcore.docs.cern.ch/development/creating-emulators/ and an example of these changes in https://github.com/dmwm/WMCore/pull/10990/files

vkuznet commented:
> @vkuznet about printing the whole error stack, how about we only do that when the component log level is set to DEBUG? Hence, we could have it in the component under a `logging.debug()` statement.
> 
> For the DBS mocking, AFAIR you have to:
> 
> 1. patch the appropriate module
> 2. create the API call signature
> 3. run a script that will really call DBS Server and it will create a huge JSON file with the content (+ the new content out of DBSErrros API)
> 
> But you can find further information under https://cms-wmcore.docs.cern.ch/development/creating-emulators/ and an example of these changes in https://github.com/dmwm/WMCore/pull/10990/files

The PR already prints full DBS error message (but not the stacktrace), see https://github.com/dmwm/WMCore/pull/12312/files#diff-dc5aa561c9a43023153ce0a627108a41a95e7bae0a9b16b474398296f23b92b7R116, and therefore there is no need for separate DEBUG option.

For mocking part I added something but I doubt anything else is needed further. Otherwise let's discuss it in private chat to speed it up.

amaltaro reviewed (changes requested):
@vkuznet as mentioned along the code, there is still a hard issue with this PR that needs to be addressed. FYI @anpicci

vkuznet commented on src/python/WMComponent/DBS3Buffer/DBSUploadPoller.py:
```diff
@@ -90,29 +92,30 @@ def uploadWorker(workInput, results, dbsUrl, gzipEncoding=False):
+            if len(records) == 1 and records[0]['block_name'] == name:
+                # found that we have this block, i.e. no need to insert block anymore
+                logging.info("block %s already exist in DBS", name)
+                results.put({'name': name, 'success': "already exist"})
```
I was not aware of this, now I adjusted this if statement to include `already exist` success value (in my view it is correct behavior and message, rather changing `already exist` to be `uploaded`). If you feel otherwise I can easily change it.

vkuznet commented on test/python/WMCore_t/Services_t/DBS_t/DBSUtils_t.py:
```diff
@@ -140,6 +142,34 @@ def testGetParallelGetParentFilesGivenParentDataset(self):
         self.assertTrue(res1 == res2)
 
+    @attr("integration")
+    def testDBSErrors(self):
```
fixed not using `DBSErrors`.

vkuznet commented on src/python/WMComponent/DBS3Buffer/DBSUploadPoller.py:
```diff
@@ -90,29 +92,30 @@ def uploadWorker(workInput, results, dbsUrl, gzipEncoding=False):
-                logging.error(msg)
-                results.put({'name': name, 'success': "error", 'error': msg})
+            meaning = dbsErrors.get(srvCode, 'DBS server does not provide code explanation')
+            msg = f'\nError trying to process block {name} through DBS. Details:'
```
done, and I updated PR description with concrete example.

amaltaro commented on src/python/WMComponent/DBS3Buffer/DBSUploadPoller.py:
```diff
@@ -775,7 +778,7 @@ def retrieveBlocks(self):
             # Remove from list of work being processed
             self.queuedBlocks.remove(result.get('name'))
-            if result["success"] == "uploaded":
+            if result["success"] == "uploaded" or result["success" == "already exist"]:
```
This line is clearly wrong. But in any case, I would stick with `uploaded` value only and keep things as simple as we can. In the end, uploaded means that the block has already been uploaded (same as already exists).

vkuznet replied:
ok, removed and use "uploaded" in both case.

vkuznet commented:
@amaltaro , I switched back to `uploaded` status as you requested and squashed the commits. Please have another look and merge if necessary.

anpicci reviewed (approved):
Thank you Valentin and Alan, I approve these changes

amaltaro reviewed (approved):
Thank you for following up on these changes, Valentin.
